service/history/api/startworkflow/api.go

```diff
@@ -41,167 +46,507 @@ import (
+		request:                    request,
+		namespace:                  namespaceEntry,
+	}, nil
+}
+
+// prepare applies request overrides, validates the request, and records eager execution metrics.
+func (s *Starter) prepare(ctx context.Context) error {
+	request := s.request.StartRequest
+	metricsHandler := s.shardCtx.GetMetricsHandler()
+
+	api.OverrideStartWorkflowExecutionRequest(request, metrics.HistoryStartWorkflowExecutionScope, s.shardCtx, metricsHandler)
+	err := api.ValidateStartWorkflowExecutionRequest(ctx, request, s.shardCtx, s.namespace, "StartWorkflowExecution")
+	if err != nil {
+		return err
+	}
+
+	if request.RequestEagerExecution {
+		metricsHandler.Counter(metrics.WorkflowEagerExecutionCounter.GetMetricName()).Record(
+			1,
+			metrics.NamespaceTag(s.namespace.Name().String()),
+			metrics.TaskQueueTag(request.TaskQueue.Name),
+			metrics.WorkflowTypeTag(request.WorkflowType.Name),
+		)
+	}
+
+	// Override to false to avoid having to look up the dynamic config throughout the diffrent code paths.
+	if !s.shardCtx.GetConfig().EnableEagerWorkflowStart(s.namespace.Name().String()) {
+		s.recordEagerDenied(eagerStartDeniedReasonDynamicConfigDisabled)
+		request.RequestEagerExecution = false
+	}
+	if s.request.FirstWorkflowTaskBackoff != nil && *s.request.FirstWorkflowTaskBackoff > 0 {
+		s.recordEagerDenied(eagerStartDeniedReasonFirstWorkflowTaskBackoff)
+		request.RequestEagerExecution = false
+	}
+	return nil
+}
+
+func (s *Starter) recordEagerDenied(reason eagerStartDeniedReason) {
+	metricsHandler := s.shardCtx.GetMetricsHandler()
+	metricsHandler.Counter(metrics.WorkflowEagerExecutionDeniedCounter.GetMetricName()).Record(
+		1,
+		metrics.NamespaceTag(s.namespace.Name().String()),
+		metrics.TaskQueueTag(s.request.StartRequest.TaskQueue.Name),
+		metrics.WorkflowTypeTag(s.request.StartRequest.WorkflowType.Name),
+		metrics.ReasonTag(string(reason)),
+	)
+}
+
+func (s *Starter) requestEagerStart() bool {
+	return s.request.StartRequest.GetRequestEagerExecution()
+}
+
+// Invoke starts a new workflow execution
+func (s *Starter) Invoke(
+	ctx context.Context,
+) (resp *historyservice.StartWorkflowExecutionResponse, retError error) {
+	request := s.request.StartRequest
+	if err := s.prepare(ctx); err != nil {
+		return nil, err
+	}
+
+	runID := uuid.NewString()
 
-	request := startRequest.StartRequest
-	api.OverrideStartWorkflowExecutionRequest(request, metrics.HistoryStartWorkflowExecutionScope, shard, shard.GetMetricsHandler())
-	err = api.ValidateStartWorkflowExecutionRequest(ctx, request, shard, namespaceEntry, "StartWorkflowExecution")
+	creationCtx, err := s.createNewMutableState(ctx, request.GetWorkflowId(), runID)
 	if err != nil {
 		return nil, err
 	}
+	err = s.createBrandNew(ctx, creationCtx)
+	if err == nil {
+		return s.generateResponse(creationCtx.runID, creationCtx.workflowTaskInfo, extractHistoryEvents(creationCtx.workflowEventBatches))
+	}
+	t, ok := err.(*persistence.CurrentWorkflowConditionFailedError)
```

Comment:
Please use Errors.As

Reply:
I copied this from the original implementation but yes, As is better here.